TypeCobol/Compiler/CupParser/NodeBuilder/ProgramClassBuilder.cs

```diff
@@ -210,6 +212,33 @@ public virtual void StartCobolCompilationUnit()
 
         public virtual void StartCobolProgram(ProgramIdentification programIdentification, LibraryCopyCodeElement libraryCopy)
         {
+            // check if a dot follows PROGRAM-ID
+            var tokens = programIdentification.ConsumedTokens.SkipWhile(t => t.TokenType != TokenType.PROGRAM_ID);
+            Token nextProgramIdToken = tokens.Skip(1).FirstOrDefault();
+            if (nextProgramIdToken != null && nextProgramIdToken.TokenType != TokenType.PeriodSeparator)
+            {
+                DiagnosticUtils.AddError(programIdentification, "Dot expected after PROGRAM-ID", nextProgramIdToken, null, MessageCode.Warning);
+            }
+
+            // check if a dot follows PROGRAM-ID declaration
+            Token nextIdentEntry = tokens.FirstOrDefault(t => t.TokenType == TokenType.AUTHOR || t.TokenType == TokenType.INSTALLATION || t.TokenType == TokenType.DATE_WRITTEN
```

Comment:
I am a little bit confused about this code, I spend around 15mn to understand why, your looking for valid tokens after the sequence of tokens: PROGRAM-ID PeriodSeparator UserDefinedWord PeriodSepartor.
So why not checking this sequence rather that having to know what are the valid tokens after the last PeriodSeparator and look back.
My Idea will be to get the index of the token PROGRAM-ID in the ConsumedTokens and check if the above sequence matches by checking Index+1, Index+2, Index+3 (Index+1 Is PeriodSeparator)? (Index+2 is UserDefinedWord)?,  (Index+3 Is PeriodSeparator)?

Reply:
Code is now in  TypeCobol/Compiler/Diagnostics/CodeElementCheckers.cs (line 324).
Ok Index+1 is PeriodSeparator (check PROGRAM-ID); but declaration can be like:
PROGRAM-ID Pgm IS INITIAL PROGRAM. (or PROGRAM-ID Pgm IS INITIAL PROGRAM)
and check (Index+3 Is PeriodSeparator) is not valid.

Reply:
Ok the grammar is not easy to parse manually.
Can you try to iterate only once on tokens?

Reply:
Use variables in grammar and check at level CobolCodeElementBuilder.
https://github.com/TypeCobolTeam/TypeCobol/pull/1651/commits/05f3a4a2e473f2087e46f6ca7e8ac2ca40ff63f3